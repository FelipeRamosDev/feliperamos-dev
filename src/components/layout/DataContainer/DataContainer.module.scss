@use '@/style/variables.scss' as *;

.DataContainer {
   display: flex;
   gap: 1rem;
   background-color: $background-dark;
   flex-direction: column;

   @media screen and (min-width: $screen-m) {
      &:not(.vertical) {
         flex-direction: row;
      }
   }

   &:not(:last-child) {
      margin-bottom: 0.7rem;
   }

   label {
      font-weight: bold;
      flex-basis: 20%;
      font-size: 0.9rem;
   }

   a {
      text-decoration: underline;
   }
}
