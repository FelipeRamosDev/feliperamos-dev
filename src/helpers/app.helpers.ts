import { CVData } from '@/types/database.types';
import { defaultLanguage } from '@/app.config';

export function apiURL(path: string, queryParams?: Record<string, string>): string {
   const baseURL = process.env.NEXT_PUBLIC_SERVER_HOST;
   const port = process.env.NEXT_PUBLIC_SERVER_HTTP_PORT;

   if (!baseURL) {
      throw new Error('NEXT_PUBLIC_SERVER_HOST is not defined');
   }

   if (!port) {
      throw new Error('NEXT_PUBLIC_SERVER_HTTP_PORT is not defined');
   }

   const url = new URL(baseURL);
   url.pathname = path;
   url.port = port;

   if (queryParams) {
      Object.entries(queryParams).forEach(([key, value]) => {
         url.searchParams.append(key, value);
      });
   }

   return url.toString();
}

export function cvPDFDownloadLink(cv: CVData, locale: string = defaultLanguage): string {
   const cvId = cv.id;
   const userFullName = cv.user?.name?.replace(/ /g, '_');

   if (!cvId || !userFullName) {
      throw new Error('CV ID or user full name is missing');
   }

   return apiURL(`/cv/${userFullName}-CV_${cvId}_${locale}.pdf`);
}

export function downloadCVPDF(cv: CVData, locale: string): void {
   if (typeof window === 'undefined') {
      return;
   }

   if (!cv) {
      throw new Error('CV data is required to download the PDF');
   }

   const link = cvPDFDownloadLink(cv, locale);
   const a = document.createElement('a');
   const userFullName = cv.user?.name;
   const cvId = cv.id;

   a.href = link;
   a.download = `${userFullName}-${cvId}(${locale}).pdf`;

   document.body.appendChild(a);
   a.click();
   document.body.removeChild(a);
}
