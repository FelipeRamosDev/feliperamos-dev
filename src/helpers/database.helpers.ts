import { FormCheckboxOption, FormSelectOption, FormValues } from "@/hooks/Form/Form.types";
import { TextResources, UserData } from "@/services";
import Ajax from "@/services/Ajax/Ajax";
import { CompanyData, EducationData, ExperienceData, LanguageData, SkillData } from "@/types/database.types";

export const handleExperienceUpdate = async (ajax: Ajax, experience: ExperienceData, values: Partial<ExperienceData>) => {
   if (!values || !Object.keys(values).length) {
      console.warn('No valid updates provided. The "updates" param is empty.');
      return { success: true };
   }

   try {
      const updated = await ajax.post<ExperienceData>('/experience/update', {
         experienceId: experience.id,
         updates: values,
      });

      if (!updated.success) {
         throw updated;
      }

      window.location.reload();
      return { success: true };
   } catch (error) {
      throw error;
   }
};

export async function loadSkillsOptions(ajax: Ajax, textResources: TextResources): Promise<FormSelectOption[]> {
   const { success, data, message } = await ajax.get<SkillData[]>('/skill/query', { params: { language_set: textResources.currentLanguage } });

   if (!success) {
      console.error('Failed to load skills:', message);
      return [];
   }

   return data.map((skill: SkillData) => ({
      value: Number(skill.id),
      label: String(skill.name),
   }));
}

export async function loadCompaniesOptions(ajax: Ajax, textResources: TextResources): Promise<FormSelectOption[]> {
   const { success, data, message } = await ajax.get<CompanyData[]>('/company/query', {
      params: { language_set: textResources.currentLanguage }
   });

   if (!success) {
      console.error('Failed to load companies:', message);
      return [];
   }

   return data.map((company: CompanyData) => ({
      value: Number(company.id),
      label: String(company.company_name)
   }));
}

export async function loadExperiencesListOptions(ajax: Ajax, language_set: string): Promise<FormCheckboxOption[]> {
   try {
      const { data = [] } = await ajax.get<ExperienceData[]>('/experience/query', { params: { language_set } });

      if (!Array.isArray(data)) {
         throw new Error("Failed to load experiences");
      }

      return data.map((item: ExperienceData) => ({
         id: item.id,
         primary: `${item.company?.company_name} (${item.position})`,
         secondary: item.title,
         avatarUrl: item.company?.logo_url
      }));
   } catch (error) {
      console.error("Error loading experiences list options:", error);
      throw error;
   }
}

export async function loadLanguagesOptions(ajax: Ajax, language_set: string): Promise<FormCheckboxOption[]> {
   try {
      const response = await ajax.get<LanguageData[]>('/user/languages', { params: { language_set } });

      if (!response.success) {
         throw new Error("Failed to load languages");
      }

      return response.data.map((item) => ({
         id: item.id,
         primary: item.default_name,
         secondary: item.level,
      }));
   } catch (error) {
      console.error("Error loading languages options:", error);
      throw error;
   }
}

export async function loadEducationsOptions(ajax: Ajax, language_set: string): Promise<FormCheckboxOption[]> {
   try {
      const response = await ajax.get<EducationData[]>('/user/educations', { params: { language_set } });

      if (!response.success) {
         throw new Error("Failed to load educations");
      }

      return response.data.map((item) => ({
         id: item.id,
         primary: item.institution_name,
         secondary: item.field_of_study,
      }));
   } catch (error) {
      console.error("Error loading educations options:", error);
      throw error;
   }
}

export async function updateUserData(ajax: Ajax, data: FormValues): Promise<UserData> {
   try {
      const updatedUser = await ajax.post<UserData>('/user/update', { updates: data });

      if (!updatedUser.success) {
         throw updatedUser;
      }

      window.location.reload();
      return updatedUser.data;
   } catch (error) {
      throw error;
   }
}
