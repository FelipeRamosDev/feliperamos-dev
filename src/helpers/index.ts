export function parseCSS(input: string | string[] = [], defaultValue: string = '') {
   if (typeof input === 'string') {
      input = [ input, defaultValue ].join(' ');
      return input;
   }

   if (Array.isArray(input)) {
      return [ ...input, defaultValue ].join(' ');
   }

   return '';
}
