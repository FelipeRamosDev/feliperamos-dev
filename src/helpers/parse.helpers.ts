export type SizeKeyword = 'xs' | 's' | 'm' | 'l' | 'xl' | 'none';
type ClassInput = string | string[] | undefined | null;

/**
 * Maps padding size keywords to corresponding CSS class names.
 *
 * @param {SizeKeyword} size - Padding size ('xs', 's', 'm', 'l', 'xl', 'none').
 * @returns {string} The CSS class name for the specified padding size, or empty string if 'none' or unrecognized.
 */
export function parsePadding(size: SizeKeyword): string {
   switch (size) {
      case 'xs':
         return 'padding-xs';
      case 's':
         return 'padding-s';
      case 'm':
         return 'padding-m';
      case 'l':
         return 'padding-l';
      case 'xl':
         return 'padding-xl';
      case 'none':
      default:
         return '';
   }
}

/**
 * Maps border-radius size keywords to corresponding CSS class names.
 *
 * @param {SizeKeyword} size - Radius size ('xs', 's', 'm', 'l', 'xl', 'none').
 * @returns {string} The CSS class name for the specified radius size, or empty string if 'none' or unrecognized.
 */
export function parseRadius(size: SizeKeyword): string {
   switch (size) {
      case 'xs':
         return 'radius-xs';
      case 's':
         return 'radius-s';
      case 'm':
         return 'radius-m';
      case 'l':
         return 'radius-l';
      case 'xl':
         return 'radius-xl';
      case 'none':
      default:
         return '';
   }
}

/**
 * Maps elevation size keywords to corresponding CSS class names.
 *
 * @param {SizeKeyword} size - Elevation size ('xs', 's', 'm', 'l', 'xl', 'none').
 * @returns {string} The CSS class name for the specified elevation size, or empty string if 'none' or unrecognized.
 */
export function parseElevation(size: SizeKeyword): string {
   switch (size) {
      case 'xs':
         return 'elevation-xs';
      case 's':
         return 'elevation-s';
      case 'm':
         return 'elevation-m';
      case 'l':
         return 'elevation-l';
      case 'xl':
         return 'elevation-xl';
      case 'none':
      default:
         return '';
   }
}

export function parseBreakpoint(size: SizeKeyword): string {
   switch (size) {
      case 'xs':
         return 'breakpoint-xs';
      case 's':
         return 'breakpoint-s';
      case 'm':
         return 'breakpoint-m';
      case 'l':
         return 'breakpoint-l';
      case 'xl':
         return 'breakpoint-xl';
      case 'none':
      default:
         return '';
   }
}

/**
 * Combines one or two sets of CSS class names into a single string.
 * Accepts strings or arrays of strings and filters out falsy values.
 *
 * @param {ClassInput} classes - First set of class names.
 * @param {ClassInput} merge - Second set of class names to merge.
 * @returns {string} Combined class names separated by spaces.
 */
export function parseCSS(classes?: ClassInput, merge?: ClassInput): string {
   let result: string[] = [];

   if (typeof classes === 'string') {
      result = [...result, ...classes.split(' ')];
   }

   if (Array.isArray(classes)) {
      result = [...result, ...classes];
   }

   if (typeof merge === 'string') {
      result = [...result, ...merge.split(' ')];
   }

   if (Array.isArray(merge)) {
      result = [...result, ...merge];
   }

   return result.filter(item => item).join(' ');
}

/**
 * Formats a numeric value as Brazilian Real currency (R$) with two decimal places.
 *
 * @param {number} value - The numeric value to format.
 * @returns {string} The formatted currency string in 'pt-BR' locale.
 */
export function parseMoney(value: number): string {
   return new Intl.NumberFormat('pt-BR', {
      style: 'currency',
      currency: 'BRL',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
   }).format(value);
}

export function parseButtonColorCSS(color?: string): string {
   if (!color || typeof color !== 'string') {
      return '';
   }

   return `button-${color}`;
}

export function onlyNumbers(value: string): string {
   return value.replace(/[^0-9]/g, '');
}
