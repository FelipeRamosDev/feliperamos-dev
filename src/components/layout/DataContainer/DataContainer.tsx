import { parseCSS, parseElevation, parsePadding, parseRadius } from '@/helpers/parse.helpers';
import styles from './DataContainer.module.scss';
import { DataContainerProps } from './DataContainer.types';

export default function DataContainer({ className, padding= 'm', radius = 's', elevation = 'none', vertical = false, children }: DataContainerProps): React.ReactElement {
   const CSS = parseCSS(className, [
      'DataContainer',
      styles.DataContainer,
      parsePadding(padding),
      parseRadius(radius),
      parseElevation(elevation),
      vertical ? styles.vertical : ''
   ]);

   return (
      <div className={CSS}>
         {children}
      </div>
   );
}
