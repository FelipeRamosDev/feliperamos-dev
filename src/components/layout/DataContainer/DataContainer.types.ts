import { SizeKeyword } from '@/helpers/parse.helpers';

export type DataContainerProps = {
   className?: string;
   padding?: SizeKeyword;
   radius?: SizeKeyword;
   elevation?: SizeKeyword;
   vertical?: boolean;
   children: React.ReactNode;
};
