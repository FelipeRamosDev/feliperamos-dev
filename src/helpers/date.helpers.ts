import { defaultLanguage } from '@/app.config';
import { DateViewStringType } from '@/types/helpers.types';
import dayjs from 'dayjs';

export function dateDifference(startDate: Date, endDate: Date, unit: 'year' | 'month' | 'day'): number {
   const _startDate = dayjs(startDate);
   const _endDate = dayjs(endDate);

   switch (unit) {
      case 'year':
         const years = _endDate.diff(_startDate, 'year');
         return years;
      case 'month':
         const months = _endDate.diff(_startDate, 'month');
         return months;
      case 'day':
         const days = _endDate.diff(_startDate, 'day');
         return days;
      default:
         return 0;
   }
}

export function dataViewString(date: Date, type: DateViewStringType = 'abbMonth-fullYear', locale: string = defaultLanguage): string {
   if (!date) {
      return '---';
   }

   switch (type) {
      case 'locale-standard':
         return dayjs(date).locale(locale).format('LL');
      case 'abbMonth-fullYear':
         return dayjs(date).locale(locale).format('MMM YYYY');
      case 'fullMonth-fullYear':
         return dayjs(date).locale(locale).format('MMMM YYYY');
      case 'fullMonth-abbYear':
         return dayjs(date).locale(locale).format('MMMM YY');
      case 'abbMonth-abbYear':
         return dayjs(date).locale(locale).format('MMM YY');
      default:
         console.warn(`Unknown date view type: ${type}`);
         return '';
   }
};

export function dateDiffYearMonth(startDate: string, endDate: string): { year: number, month: number } {
   const years = dateDifference(new Date(startDate), new Date(endDate), 'year');
   const months = dateDifference(new Date(startDate), new Date(endDate), 'month');
   const restMonths = months % 12;

   return {
      year: years,
      month: restMonths
   };
}

