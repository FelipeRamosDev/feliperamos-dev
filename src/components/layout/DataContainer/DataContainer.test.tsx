import React from 'react';
import { render, screen } from '@testing-library/react';
import DataContainer from './DataContainer';
import { DataContainerProps } from './DataContainer.types';
import { SizeKeyword } from '@/helpers/parse.helpers';

// Mock the parse helpers
jest.mock('@/helpers/parse.helpers', () => ({
   parseCSS: jest.fn(() => 'mocked-parse-css'),
   parseElevation: jest.fn(() => 'mocked-elevation'),
   parsePadding: jest.fn(() => 'mocked-padding'),
   parseRadius: jest.fn(() => 'mocked-radius'),
}));

// Mock the SCSS module
jest.mock('../DataContainer.module.scss', () => ({
   DataContainer: 'mocked-container-class',
   vertical: 'mocked-vertical-class',
}));

// Import mocked functions for type checking
import { parseCSS, parseElevation, parsePadding, parseRadius } from '@/helpers/parse.helpers';

const mockParseCSS = parseCSS as jest.MockedFunction<typeof parseCSS>;
const mockParseElevation = parseElevation as jest.MockedFunction<typeof parseElevation>;
const mockParsePadding = parsePadding as jest.MockedFunction<typeof parsePadding>;
const mockParseRadius = parseRadius as jest.MockedFunction<typeof parseRadius>;

describe('DataContainer', () => {
   beforeEach(() => {
      jest.clearAllMocks();
      // Reset default return values
      mockParseCSS.mockReturnValue('mocked-parse-css');
      mockParseElevation.mockReturnValue('mocked-elevation');
      mockParsePadding.mockReturnValue('mocked-padding');
      mockParseRadius.mockReturnValue('mocked-radius');
   });

   const defaultProps: DataContainerProps = {
      children: <div>Test content</div>,
   };

   describe('Basic Rendering', () => {
      it('renders without crashing', () => {
         render(<DataContainer {...defaultProps} />);
         expect(screen.getByText('Test content')).toBeInTheDocument();
      });

      it('renders children correctly', () => {
         render(
            <DataContainer {...defaultProps}>
               <span>Child 1</span>
               <span>Child 2</span>
            </DataContainer>
         );
         
         expect(screen.getByText('Child 1')).toBeInTheDocument();
         expect(screen.getByText('Child 2')).toBeInTheDocument();
      });

      it('renders with div container element', () => {
         const { container } = render(<DataContainer {...defaultProps} />);
         const dataContainer = container.firstChild as HTMLElement;
         
         expect(dataContainer.tagName).toBe('DIV');
      });
   });

   describe('Props Handling', () => {
      describe('className prop', () => {
         it('applies custom className when provided', () => {
            const { container } = render(
               <DataContainer {...defaultProps} className="custom-class" />
            );
            const dataContainer = container.firstChild as HTMLElement;
            
            expect(mockParseCSS).toHaveBeenCalledWith(
               'custom-class',
               expect.arrayContaining([
                  'DataContainer',
                  'mocked-container-class',
                  'mocked-padding',
                  'mocked-radius',
                  'mocked-elevation',
                  ''
               ])
            );
            expect(dataContainer).toHaveClass('mocked-parse-css');
         });

         it('does not apply custom className when not provided', () => {
            const { container } = render(<DataContainer {...defaultProps} />);
            const dataContainer = container.firstChild as HTMLElement;
            
            expect(dataContainer.className).not.toContain('custom-class');
         });
      });

      describe('padding prop', () => {
         it('calls parsePadding with default value "m" when not provided', () => {
            render(<DataContainer {...defaultProps} />);
            
            expect(mockParsePadding).toHaveBeenCalledWith('m');
         });

         it('calls parsePadding with provided value', () => {
            render(<DataContainer {...defaultProps} padding="l" />);
            
            expect(mockParsePadding).toHaveBeenCalledWith('l');
         });

         it('calls parsePadding with different size values', () => {
            const sizes: SizeKeyword[] = ['xs', 's', 'm', 'l', 'xl'];
            
            sizes.forEach(size => {
               render(<DataContainer {...defaultProps} padding={size} />);
               expect(mockParsePadding).toHaveBeenCalledWith(size);
            });
         });
      });

      describe('radius prop', () => {
         it('calls parseRadius with default value "s" when not provided', () => {
            render(<DataContainer {...defaultProps} />);
            
            expect(mockParseRadius).toHaveBeenCalledWith('s');
         });

         it('calls parseRadius with provided value', () => {
            render(<DataContainer {...defaultProps} radius="m" />);
            
            expect(mockParseRadius).toHaveBeenCalledWith('m');
         });

         it('calls parseRadius with different size values', () => {
            const sizes: SizeKeyword[] = ['xs', 's', 'm', 'l', 'xl'];
            
            sizes.forEach(size => {
               render(<DataContainer {...defaultProps} radius={size} />);
               expect(mockParseRadius).toHaveBeenCalledWith(size);
            });
         });
      });

      describe('elevation prop', () => {
         it('calls parseElevation with default value "none" when not provided', () => {
            render(<DataContainer {...defaultProps} />);
            
            expect(mockParseElevation).toHaveBeenCalledWith('none');
         });

         it('calls parseElevation with provided value', () => {
            render(<DataContainer {...defaultProps} elevation="m" />);
            
            expect(mockParseElevation).toHaveBeenCalledWith('m');
         });

         it('calls parseElevation with different size values', () => {
            const sizes: SizeKeyword[] = ['none', 'xs', 's', 'm', 'l', 'xl'];
            
            sizes.forEach(size => {
               render(<DataContainer {...defaultProps} elevation={size} />);
               expect(mockParseElevation).toHaveBeenCalledWith(size);
            });
         });
      });

      describe('vertical prop', () => {
         it('applies vertical class when vertical is true', () => {
            const { container } = render(
               <DataContainer {...defaultProps} vertical={true} />
            );
            const dataContainer = container.firstChild as HTMLElement;
            
            expect(mockParseCSS).toHaveBeenCalledWith(
               undefined,
               expect.arrayContaining([
                  'DataContainer',
                  'mocked-container-class',
                  'mocked-padding',
                  'mocked-radius',
                  'mocked-elevation',
                  'mocked-vertical-class'
               ])
            );
            expect(dataContainer).toHaveClass('mocked-parse-css');
         });

         it('does not apply vertical class when vertical is false', () => {
            const { container } = render(
               <DataContainer {...defaultProps} vertical={false} />
            );
            const dataContainer = container.firstChild as HTMLElement;
            
            expect(dataContainer).not.toHaveClass('mocked-vertical-class');
         });

         it('does not apply vertical class when vertical is not provided (default false)', () => {
            const { container } = render(<DataContainer {...defaultProps} />);
            const dataContainer = container.firstChild as HTMLElement;
            
            expect(dataContainer).not.toHaveClass('mocked-vertical-class');
         });
      });
   });

   describe('CSS Class Management', () => {
      it('calls parseCSS with correct parameters', () => {
         render(
            <DataContainer 
               {...defaultProps}
               className="custom-class"
               padding="l"
               radius="m"
               elevation="s"
               vertical={true}
            />
         );
         
         expect(mockParseCSS).toHaveBeenCalledWith(
            'custom-class',
            [
               'DataContainer',
               'mocked-container-class',
               'mocked-padding',
               'mocked-radius',
               'mocked-elevation',
               'mocked-vertical-class'
            ]
         );
      });

      it('calls parseCSS without custom className', () => {
         render(
            <DataContainer 
               {...defaultProps}
               padding="s"
               radius="l"
               elevation="m"
               vertical={false}
            />
         );
         
         expect(mockParseCSS).toHaveBeenCalledWith(
            undefined,
            [
               'DataContainer',
               'mocked-container-class',
               'mocked-padding',
               'mocked-radius',
               'mocked-elevation',
               ''
            ]
         );
      });

      it('applies base container class', () => {
         const { container } = render(<DataContainer {...defaultProps} />);
         const dataContainer = container.firstChild as HTMLElement;
         
         expect(dataContainer).toHaveClass('mocked-parse-css');
      });

      it('combines all classes correctly', () => {
         render(
            <DataContainer 
               {...defaultProps}
               className="test-class"
               vertical={true}
            />
         );
         
         const expectedClasses = [
            'DataContainer',
            'mocked-container-class',
            'mocked-padding',
            'mocked-radius',
            'mocked-elevation',
            'mocked-vertical-class'
         ];
         expect(mockParseCSS).toHaveBeenCalledWith('test-class', expectedClasses);
      });
   });

   describe('Children Rendering', () => {
      it('renders single child element', () => {
         render(
            <DataContainer {...defaultProps}>
               <button>Click me</button>
            </DataContainer>
         );
         
         expect(screen.getByRole('button', { name: 'Click me' })).toBeInTheDocument();
      });

      it('renders multiple child elements', () => {
         render(
            <DataContainer {...defaultProps}>
               <h1>Title</h1>
               <p>Paragraph</p>
               <button>Button</button>
            </DataContainer>
         );
         
         expect(screen.getByRole('heading', { name: 'Title' })).toBeInTheDocument();
         expect(screen.getByText('Paragraph')).toBeInTheDocument();
         expect(screen.getByRole('button', { name: 'Button' })).toBeInTheDocument();
      });

      it('renders nested components', () => {
         const NestedComponent = () => <span>Nested content</span>;
         
         render(
            <DataContainer {...defaultProps}>
               <div>
                  <NestedComponent />
               </div>
            </DataContainer>
         );
         
         expect(screen.getByText('Nested content')).toBeInTheDocument();
      });

      it('renders text nodes', () => {
         render(
            <DataContainer {...defaultProps}>
               Just plain text
            </DataContainer>
         );
         
         expect(screen.getByText('Just plain text')).toBeInTheDocument();
      });
   });

   describe('Accessibility', () => {
      it('maintains proper document structure', () => {
         const { container } = render(
            <DataContainer {...defaultProps}>
               <h1>Main heading</h1>
               <p>Content paragraph</p>
            </DataContainer>
         );
         
         const heading = screen.getByRole('heading', { level: 1 });
         const paragraph = screen.getByText('Content paragraph');
         
         expect(heading).toBeInTheDocument();
         expect(paragraph).toBeInTheDocument();
         expect(container.firstChild).toContainElement(heading);
         expect(container.firstChild).toContainElement(paragraph);
      });

      it('does not interfere with child accessibility features', () => {
         render(
            <DataContainer {...defaultProps}>
               <button aria-label="Close dialog">×</button>
               <div role="alert">Error message</div>
            </DataContainer>
         );
         
         expect(screen.getByLabelText('Close dialog')).toBeInTheDocument();
         expect(screen.getByRole('alert')).toBeInTheDocument();
      });

      it('preserves semantic markup', () => {
         render(
            <DataContainer {...defaultProps}>
               <main>
                  <section>
                     <article>Content</article>
                  </section>
               </main>
            </DataContainer>
         );
         
         expect(screen.getByRole('main')).toBeInTheDocument();
         expect(screen.getByText('Content')).toBeInTheDocument();
      });
   });

   describe('Edge Cases', () => {
      it('handles empty children gracefully', () => {
         render(<DataContainer>{undefined}</DataContainer>);
         
         const { container } = render(<DataContainer>{undefined}</DataContainer>);
         expect(container.firstChild).toBeInTheDocument();
      });

      it('handles null children', () => {
         render(<DataContainer>{null}</DataContainer>);
         
         const { container } = render(<DataContainer>{null}</DataContainer>);
         expect(container.firstChild).toBeInTheDocument();
      });

      it('handles conditional children', () => {
         const showContent = true;
         render(
            <DataContainer>
               {showContent && <div>Conditional content</div>}
            </DataContainer>
         );
         
         expect(screen.getByText('Conditional content')).toBeInTheDocument();
      });

      it('handles array of children', () => {
         const items = ['Item 1', 'Item 2', 'Item 3'];
         render(
            <DataContainer>
               {items.map((item, index) => (
                  <div key={index}>{item}</div>
               ))}
            </DataContainer>
         );
         
         items.forEach(item => {
            expect(screen.getByText(item)).toBeInTheDocument();
         });
      });

      it('handles complex nested structures', () => {
         render(
            <DataContainer vertical={true} className="complex-container">
               <header>
                  <h1>Header</h1>
               </header>
               <main>
                  <section>
                     <article>
                        <p>Article content</p>
                     </article>
                  </section>
               </main>
               <footer>
                  <p>Footer</p>
               </footer>
            </DataContainer>
         );
         
         expect(screen.getByRole('banner')).toBeInTheDocument();
         expect(screen.getByRole('main')).toBeInTheDocument();
         expect(screen.getByRole('contentinfo')).toBeInTheDocument();
         expect(screen.getByText('Article content')).toBeInTheDocument();
      });

      it('handles all props together', () => {
         render(
            <DataContainer
               className="all-props-test"
               padding="xl"
               radius="l"
               elevation="m"
               vertical={true}
            >
               <div>All props applied</div>
            </DataContainer>
         );
         
         expect(mockParseCSS).toHaveBeenCalledWith(
            'all-props-test',
            [
               'DataContainer',
               'mocked-container-class',
               'mocked-padding',
               'mocked-radius',
               'mocked-elevation',
               'mocked-vertical-class'
            ]
         );
         expect(mockParsePadding).toHaveBeenCalledWith('xl');
         expect(mockParseRadius).toHaveBeenCalledWith('l');
         expect(mockParseElevation).toHaveBeenCalledWith('m');
         expect(screen.getByText('All props applied')).toBeInTheDocument();
      });
   });

   describe('Snapshots', () => {
      it('matches snapshot with default props', () => {
         const { container } = render(
            <DataContainer>
               <div>Default snapshot content</div>
            </DataContainer>
         );
         
         expect(container.firstChild).toMatchSnapshot();
      });

      it('matches snapshot with all props', () => {
         const { container } = render(
            <DataContainer
               className="snapshot-test"
               padding="l"
               radius="m"
               elevation="s"
               vertical={true}
            >
               <div>Full props snapshot</div>
            </DataContainer>
         );
         
         expect(container.firstChild).toMatchSnapshot();
      });

      it('matches snapshot with complex children', () => {
         const { container } = render(
            <DataContainer vertical={true}>
               <header>
                  <h1>Complex Header</h1>
               </header>
               <main>
                  <p>Main content</p>
                  <button>Action</button>
               </main>
            </DataContainer>
         );
         
         expect(container.firstChild).toMatchSnapshot();
      });
   });
});
