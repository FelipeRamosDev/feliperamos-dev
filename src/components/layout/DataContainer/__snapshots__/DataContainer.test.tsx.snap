// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataContainer Snapshots matches snapshot with all props 1`] = `
<div
  class="mocked-parse-css"
>
  <div>
    Full props snapshot
  </div>
</div>
`;

exports[`DataContainer Snapshots matches snapshot with complex children 1`] = `
<div
  class="mocked-parse-css"
>
  <header>
    <h1>
      Complex Header
    </h1>
  </header>
  <main>
    <p>
      Main content
    </p>
    <button>
      Action
    </button>
  </main>
</div>
`;

exports[`DataContainer Snapshots matches snapshot with default props 1`] = `
<div
  class="mocked-parse-css"
>
  <div>
    Default snapshot content
  </div>
</div>
`;
